### PR TITLE
Add a "create test" icon to test editors

### PR DESCRIPTION
**Pull request by johnpaliotta**
This PR makes the following changes

- Allow creation of VectorCAST tests from test editors.
A "flask" icon is displayed for each function or method that is testable.  To create a new test, right click on the icon and choose: "Add VectorCAST Test." 
- Fix issue #9 
- Fix issue #11 
- Fix a test tree update issue caused by the VS Code 1.81 release, that caused the test tree to not update after a load or delete action.
- Log calls to vpython scripts when "Verbose Logging" option is enabled
- Some naming improvements

### CODE REVIEW

File: src/utilities.ts [hunk 1]
```diff
@@ -535,3 +538,32 @@ export function checkIfInstallationIsOK() {
...
+			if (path.charAt(1) == ":") {
+				const driveLetter = path.charAt(0).toUpperCase();
+				return driveLetter + path.slice(1, path.length);
+			}
+		}
+		return path;
+		}
+	else
+		return ""
+}
```
Comment (aytey) [inline src/utilities.ts:561]: Can we run prettier on this, please?

File: README.md [hunk 2]
```diff
@@ -166,18 +166,19 @@ This extension contributes the following settings:
 - "Show Report on Execute" will show the HTML test report in after each VectorCAST test run.
 - "Decorate Explorer Tree" will add a VC in the right margin of the file explorer tree for those files that have VectorCAST coverage.
 - "Unit Test Location" controls where the extension stores unit test new unit test artifacts.
-- "Verbose Logging" will add more detailed messages to the VectorCAST Test Explorer message pane
+- "Verbose Logging" will add more detailed messages to the VectorCAST Test Explorer message panel
 
 ## Known Issues
 
-- If a VectorCAST configuration file (CCAST\_.CFG) does not exist in the directory chosen
+- If a VectorCAST configuration file (CCAST_.CFG) does not exist in the directory chosen
```
Comment (aytey) [inline README.md:173]: The escape was probably correct; it stops Markdown thinking you're starting an italics block (between two `_`)

File: src/testPane.ts [hunk 3]
```diff
@@ -362,9 +368,9 @@ export function updateTestsForEnvironment(
   let jsonData = getEnviroDataFromPython(enviroPath);
 
   if (jsonData) {
-    let nodeID: string;
+    let nodeID: string = "vcast:"
```
Comment (aytey) [inline src/testPane.ts:371]: Is this for a user or for us? What's the standard for other extensions?

File: src/testPane.ts [hunk 4]
```diff
@@ -386,6 +392,7 @@ export function updateTestsForEnvironment(
         vcastEnviroList
       );
     }
+    controller.items.delete(enviroNode.id);
```
Comment (aytey) [inline src/testPane.ts:395]: We should add a comment about why this delete is required; reading the docs for `add` it _should not_ be required.

File: src/testPane.ts [hunk 5]
```diff
@@ -447,8 +454,10 @@ export async function loadAllVCTests(
     } // for workspace folders
   } // if workspace folders
 
-  // once the data is loaded update the coverage for the active editor
+  // once the data is loaded update the coverage and test icons for the active editor
   updateDisplayedCoverage();
+  updateTestDecorator ();
```
Comment (aytey) [inline src/testPane.ts:459]: Prettier to fix the space

File: python/vTestInterface.py [hunk 6]
```diff
@@ -254,6 +255,20 @@ def getUnitData(enviroPath, kind):
     return unitList
 
 
+def getFunctionData (soruceObject):
```
Comment (aytey) [inline python/vTestInterface.py:258]: Spelling of `soruceObject`

File: src/editorDecorator.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+export function updateFunctionDataForFile(
+    enviroPath:string,
+    fileName: string,
+    functionList: string[]) {
+
+    // functionList is a list of json items with fields for "name" and "startLine"
+    let lineMap:Map<number, string> = new Map();
+    for (let i = 0; i < functionList.length; i++) {
+        const functionInfo:any = functionList[i];
```
Comment (aytey) [inline src/editorDecorator.ts:55]: Useless type annotation

File: CHANGELOG.md [hunk 8]
```diff
@@ -31,4 +31,16 @@ by the time clicast was called to load the script.
...
-### Fixed spelling errors / typos in source code comments
\ No newline at end of file
+### Fixed spelling errors / typos in source code comments
+
+
+## [1.0.4] - 2023-08-22
+
+### Allow creation of VectorCAST tests from test editors
+- A "flask" icon is displayed for each function or method that is testable. 
+- To create a new test, right click on the icon and choose: "Add VectorCAST Test"
```
[the remark this answers is not in the document]
Comment (aytey) [reply CHANGELOG.md:41]: Actually, it appears we allow someone to click where the flask _was_ in any file:

image

File: src/vcastTestInterface.ts [hunk 9]
```diff
@@ -228,14 +235,14 @@ export function getListOfFilesWithCoverage(): string[] {
...
   enviroPath: string,
   fileList: any[]
 ) {
   let filePathList: string[] = [];
 
   for (let fileIndex = 0; fileIndex < fileList.length; fileIndex++) {
-    let filePath = fileList[fileIndex].path;
+    let filePath = fixDriveLetterForPython (fileList[fileIndex].path);
 
     // Improvement needed: should make a function for this
```
[the remark this answers is not in the document]
Comment (johnpaliotta) [reply src/vcastTestInterface.ts:248]: Done

File: README.md (hunk 2 above)
Comment (johnpaliotta) [reply README.md:173]: Ahh ok.  It renders fine either way though ...

File: src/testPane.ts (hunk 3 above)
Comment (johnpaliotta) [reply src/testPane.ts:371]: No one sees this, it is an internal id.
The other extensions that I looked at do not add context menus, so this was a bit "uncharted" so I came up with this really easy fix that leverages the "when clause" regex feature - see package.json.

File: src/testPane.ts (hunk 4 above)
Comment (johnpaliotta) [reply src/testPane.ts:395]: Done.
I am planning to create an issue against vs code also

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnpaliotta) [reply CHANGELOG.md:41]: Wow, good catch on this one ... thanks!

File: python/vTestInterface.py (hunk 6 above)
Comment (johnpaliotta) [reply python/vTestInterface.py:258]: Done

File: src/utilities.ts (hunk 1 above)
Comment (johnpaliotta) [reply src/utilities.ts:561]: Done

File: src/editorDecorator.ts (hunk 7 above)
Comment (aytey) [reply src/editorDecorator.ts:55]: You've marked a variable as type `:any`, which just means "this is any type" (so you might as well not have it)

File: src/testPane.ts (hunk 3 above)
Comment (aytey) [reply src/testPane.ts:371]: > No one sees this, it is an internal id.
>

Well, we sort of do:

image

File: src/testPane.ts (hunk 4 above)
Comment (aytey) [reply src/testPane.ts:396]: I've already asked @Zbigor to try to create a reproducer for this.

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (aytey) [reply CHANGELOG.md:41]: Appears fixed; I can no longer reproduce this.

File: src/testPane.ts (hunk 3 above)
Comment (johnpaliotta) [reply src/testPane.ts:371]: Should have said, no one SHOULD see this

Comment (johnpaliotta) [reply src/testPane.ts:371]: I did not notice this because my directory name is vcastTutorial :)

Comment (johnpaliotta) [reply src/testPane.ts]: Fixed

File: src/editorDecorator.ts (hunk 7 above)
Comment (aytey) [reply src/editorDecorator.ts:55]: Wow, fair enough.

File: src/utilities.ts [hunk 10]
```diff
@@ -535,3 +538,32 @@ export function checkIfInstallationIsOK() {
...
+    if (platform == "win32") {
+      if (path.charAt(1) == ":") {
+        const driveLetter = path.charAt(0).toLowerCase();
+        return driveLetter + path.slice(1, path.length);
+      }
+    }
+    return path;
+  }
+  else
+    return ""
```
Comment (aytey) [inline src/utilities.ts:560]: Why this return and not just `path`?

File: src/vcastTestInterface.ts [hunk 11]
```diff
@@ -722,12 +727,13 @@ function createScriptTemplate(testNode: testNodeType): string {
   return scriptTemplateLines.join("\n");
 }
 
-export async function newTestScript(nodeID: string) {
+
+
+export async function newTestScript(testNode: testNodeType) {
```
Comment (aytey) [inline src/vcastTestInterface.ts:732]: Very minor, but I am not sure why we added spaces here (I'm also surprised `prettier` didn't remove them).

**Review by aytey — APPROVED** [review]
LGTM

**Pull request merged**